src/base/Identifier.ml

```diff
@@ -19,21 +19,89 @@
 open Core_kernel
 open ErrorUtils
 
-type 'rep t = Ident of string * 'rep [@@deriving sexp]
+module type QualifiedName = sig
 
-let asId i = Ident (i, dummy_loc)
+  type t [@@deriving sexp]
 
-let asIdL i loc = Ident (i, loc)
+  val as_string : t -> string
 
-let get_id i = match i with Ident (x, _) -> x
+  val as_error_string : t -> string
 
-let get_rep i = match i with Ident (_, l) -> l
+  val equal_id : t -> t -> bool
+  
+  val compare_id : t -> t -> int
+    
+end
+
+module FlattenedName = struct
+
+  type t = string [@@deriving sexp]
+
+  let as_string n = n
+
+  let as_error_string n = n
+
+  let equal_id = String.(=)
+
+  let compare_id = String.compare
+
+end
+
+
+module type ScillaIdentifier = sig
+
+  module Name : QualifiedName
+    
+  type 'rep t = Ident of Name.t * 'rep [@@deriving sexp]
```

Comment:
What if we made type `t` abstract? We have both constructors and destructors below. Or, if we want to still keep pattern-matching, we could declare it like so
```
type 'rep t = private Ident of Name.t * 'rep [@@deriving sexp]
```
This prevents constructing new terms of type `t` using `Ident` constructor, but allows the user to destruct.

If we decide to take this path, then it looks like functions like these could be useful
```
val annotate : 'a t -> 'b -> ('a * 'b) t
val map1 : ('a * 'b) t -> ('a -> 'c) -> ('c * 'b) t
...
```

Reply:
Won't the `Ident` constructor be hidden when the module is restricted by `: QualifiedName`?

in any case we might as well use `get_id` and `get_rep` to access the components of `Ident`, so I don't see the value of making `Ident` available for destruction either.

Reply:
Oh, sorry I meant this comment to go to the signature, not structure.

> I don't see the value of making Ident available for destruction either.

It's just sometimes it is less verbose to name all the projections at once as opposed to write them one-by-one, multiplicating the name of the term being destructed.

Reply:
Hmm... I see what you mean. I like the way it's currently wrapped in constructor and destructor functions, but that's only nice when there is only one constructor. I don't see us adding more constructors, but I suppose we might as well make the constructor available as well.